src/cmd/factory/impl/common.go

```diff
@@ -179,3 +199,547 @@ func buildCollections(
 
 	return collections, nil
 }
+
+type permData struct {
+	user        string // user is only for older versions
+	entityID    string
+	roles       []string
+	sharingMode onedrive.SharingMode
+}
+
+type itemData struct {
+	name  string
+	data  []byte
+	perms permData
+}
+
+type itemInfo struct {
+	// lookupKey is a string that can be used to find this data from a set of
+	// other data in the same collection. This key should be something that will
+	// be the same before and after restoring the item in M365 and may not be
+	// the M365 ID. When restoring items out of place, the item is assigned a
+	// new ID making it unsuitable for a lookup key.
+	lookupKey string
+	name      string
+	data      []byte
+}
+
+type onedriveCollection struct {
+	service       path.ServiceType
+	pathElements  []string
+	items         []itemInfo
+	aux           []itemInfo
+	backupVersion int
+}
+
+type onedriveColInfo struct {
+	pathElements []string
+	perms        permData
+	files        []itemData
+	folders      []itemData
+}
+
+var (
+	folderAName = "folder-a"
+	folderBName = "b"
+	folderCName = "folder-c"
+
+	fileAData = []byte(strings.Repeat("a", 33))
+	fileBData = []byte(strings.Repeat("b", 65))
+	fileEData = []byte(strings.Repeat("e", 257))
+
+	// Cannot restore owner or empty permissions and so not testing them
+	writePerm = []string{"write"}
+	readPerm  = []string{"read"}
+)
+
+func generateAndRestoreOnedriveItems(
+	gc *connector.GraphConnector,
+	resourceOwner, secondaryUserID, secondaryUserName string,
+	acct account.Account,
+	service path.ServiceType,
+	cat path.CategoryType,
+	sel selectors.Selector,
+	tenantID, destFldr string,
+	count int,
+	errs *fault.Bus,
+) (
+	*details.Details,
+	error,
+) {
+	ctx, flush := tester.NewContext()
+	defer flush()
+
+	// TODO: fit the destination to the containers
+	dest := control.DefaultRestoreDestination(common.SimpleTimeTesting)
+	dest.ContainerName = destFldr
+	print.Infof(ctx, "Restoring to folder %s", dest.ContainerName)
+
+	d, _ := gc.Service.Client().UsersById(resourceOwner).Drive().Get(ctx, nil)
+	driveID := ptr.Val(d.GetId())
+
+	var (
+		cols []onedriveColInfo
+
+		rootPath = []string{
+			"drives",
+			driveID,
+			"root:",
+		}
```

Comment:
nit, this isn't a very long line
```suggestion
		rootPath = []string{ "drives", driveID, "root:"}
```